compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/generators/AdapterGenerator.kt

```diff
@@ -448,19 +452,41 @@ internal class AdapterGenerator(
         }
     }
 
-    /**
-     * For example,
-     * fun consumer(f: suspend () -> Unit) = ...
-     * fun nonSuspendFunction = { ... }
-     * fun useSite(...) = { ... consumer(nonSuspendFunction) ... }
-     *
-     * At the use site, instead of the argument, we can put the suspend lambda as an adapter.
-     */
-    private fun needSuspendConversion(argument: FirExpression, parameter: FirValueParameter): Boolean =
-        // TODO: should refer to LanguageVersionSettings.SuspendConversion
-        parameter.returnTypeRef.coneType.isSuspendFunctionType(session) &&
-                argument.typeRef.coneType.isBuiltinFunctionalType(session) &&
-                !argument.typeRef.coneType.isSuspendFunctionType(session)
+    private fun findInvokeSymbol(type: IrSimpleType, argument: FirExpression): IrSimpleFunctionSymbol? {
+        val argumentType = argument.typeRef.coneType
+        // Either the argument type is a non-suspend functional type
+        if (argumentType.isBuiltinFunctionalType(session) && !argumentType.isSuspendFunctionType(session)) {
+            return (argument.typeRef.coneType as? ConeClassLikeType)?.lookupTag?.classId
+                ?.let { classId ->
+                    session.firSymbolProvider
+                        .getClassDeclaredCallableSymbols(classId, Name.identifier("invoke"))
+                        .filterIsInstance<FirFunctionSymbol<*>>()
+                        .find { firFunctionSymbol ->
+                            firFunctionSymbol.fir.valueParameters.size == type.arguments.size - 1
+                        }?.let { firFunctionSymbol ->
+                            declarationStorage.getIrFunctionSymbol(firFunctionSymbol) as? IrSimpleFunctionSymbol
+                        }
+                }
+        }
+
+        // Or is a class-like type whose corresponding class is a user-defined and has an overridden `invoke`.
```

Comment:
I'm concerned about this part.
AFAIK, suspend conversion works only on a value of a function type `(P1,.., PN) -> R` that is being used as a value for `suspend (P1,.., PN) -> R`, i.e. it shouldn't for an arbitrary class `X` that has an invoke irrelevant to one from function type.
So, I'd check that current `argumentType` is a subtype of function type and then looked for a symbol into a scope instead of declarations because the symbol might be inherited from supertypes, for example

Reply:
I should have paid attention to the test name, which itself implies that subtype of functional type is applied to suspend conversion. :o It looks like a user class can implement multiple functional types, so the expected number of type parameter should be accounted as well.

The 2nd commit, https://github.com/JetBrains/kotlin/pull/3718/commits/ebe92c7a6457d21156dca30916b96d1dd6e4e066, addressed the comment, including the use of `scope`.